web/sites/default/settings.php

```diff
@@ -2,264 +2,80 @@
 
-    // Manually add the classloader path, this is required for the container cache bin definition below
-    // and allows to use it without the redis module being enabled.
-    $class_loader->addPsr4('Drupal\\redis\\', 'modules/contrib/redis/src');
+// Todo, this logic to be reviewed by Amazee.
+$govcms_is_prod = !getenv('DEV_MODE') && (getenv('LAGOON_ENVIRONMENT_TYPE') && getenv('LAGOON_ENVIRONMENT_TYPE') == 'production');
 
-    // Use redis for container cache.
-    // The container cache is used to load the container definition itself, and
-    // thus any configuration stored in the container itself is not available
-    // yet. These lines force the container cache to use Redis rather than the
-    // default SQL cache.
-    $settings['bootstrap_container_definition'] = [
-      'parameters' => [],
-      'services' => [
-        'redis.factory' => [
-          'class' => 'Drupal\redis\ClientFactory',
-        ],
-        'cache.backend.redis' => [
-          'class' => 'Drupal\redis\Cache\CacheBackendFactory',
-          'arguments' => ['@redis.factory', '@cache_tags_provider.container', '@serialization.phpserialize'],
-        ],
-        'cache.container' => [
-          'class' => '\Drupal\redis\Cache\PhpRedis',
-          'factory' => ['@cache.backend.redis', 'get'],
-          'arguments' => ['container'],
-        ],
-        'cache_tags_provider.container' => [
-          'class' => 'Drupal\redis\Cache\RedisCacheTagsChecksum',
-          'arguments' => ['@redis.factory'],
-        ],
-        'serialization.phpserialize' => [
-          'class' => 'Drupal\Component\Serialization\PhpSerialize',
-        ],
-      ],
-    ];
-  }
+if (!file_exists($govcms_includes . '/all.settings.php')) {
+  // Currently simesy/require.
+  throw new \Exception('Expecting settings files at ' . $govcms_includes . '. Has "composer install" run successfully?');
 }
 
-// Public, private and temporary files paths.
-if (getenv('LAGOON')) {
-  $settings['file_public_path'] = 'sites/default/files';
-  $settings['file_private_path'] = 'sites/default/files/private';
-  $settings['file_temporary_path'] = 'sites/default/files/private/tmp';
-}
+include $govcms_includes . '/all.settings.php';
 
-// ClamAV settings.
-$config['clamav.settings']['scan_mode'] = 1;
-$config['clamav.settings']['mode_executable']['executable_path'] = '/usr/bin/clamscan';
-
-// Enable shield.
-if (getenv('LAGOON') && (getenv('DRUPAL_SHIELD_USER') && getenv('DRUPAL_SHIELD_PASS'))) {
-  $config['shield.settings']['credential_provider'] = 'shield';
-  $config['shield.settings']['credentials']['shield']['user'] = getenv('DRUPAL_SHIELD_USER');
-  $config['shield.settings']['credentials']['shield']['pass'] = getenv('DRUPAL_SHIELD_PASS');
-}
-
-
-// Hash Salt
+// Every Lagoon, local or remote.
 if (getenv('LAGOON')) {
-  $settings['hash_salt'] = hash('sha256', getenv('LAGOON_PROJECT'));
+  include $govcms_includes . '/lagoon.settings.php';
```

Comment:
can all this logic be moved to the tooling repo? The settings.php file will then only have:
- inclusion of the settings from the tooling
- including of the `settings.local.php` file (how default settings.php has it).

This way, it all will be much cleaner

Reply:
We discussed this today. I've argued that having some procedural style logic here is much more transparent to the developer who is exploring it. But I think this is still "suck and see".

Reply:
For example a developer could var_dump() after each of these statements to see what is being set at each point. They can't push these changes to the platform, but this is easier for many developers than changing files in vendor (which is more scary for new devs).

Reply:
They should use xdebug sure. That's where the community help/docs will come in. :)

Reply:
so right now the inclusions are:
- all settings
- lagoon
- per environment
- per project
- local

The problem with this (apart that it is too hard to understand the purpuse of each one) is that you still have settings stored outside of the Git index. Previously we had all these files sitting in git, and , while still having a lot of them, at least it was possible to see what is happening inside of them. Now it is mix of files in Git and outside of it. Nothing worse than magically included files.

But even with these inclusions, some of the files are now stored outside of Git index, but at the same time included in this settings file explicitly, which is even more confusing: why do I as a developer now have an option to remove these lines? Should'nt production settings be locked?

Reply:
As a SaaS developer you can't remove any/override *any* of this and you will just edit settings.local.php. That's a *really* common pattern.

As a PaaS developer you can knock yourself out - replace settings.php and do whatever you like. I'll have my ears open for other opinions on this...

Including settings config outside git is common across most platforms - acquia and platform have techniques. By removing all this common stuff form the project repo we avoid drift, while still allowing developers to override stuff easily (for the whole project if permitted) or just locally.

The problem we need to solve is that currently all the settings files are being completely ignored and new settings are all dumped into settings.php. There are also *in some cases* settings files being injected from govcms8lagoon image.

So really all this does is I've just reverted to the Lagoon suggested structure, with more relevant commenting in the settings.php (see the developer stories at the top of the file) so that future PRs we can point to these instructions. And move them out of the repo to stop the settings rot.

Reply:
Please refer to https://github.com/govCMS/govcms8lagoon/tree/master/.docker/images/govcms8/settings which is existing configuration we have outside of the repo. I'm intending to remove these files and enforcing a standard pattern.

Reply:
saas can't edit anything - no change.
local.settings.php for local - common pattern.
PaaS can do anything - see comments at top of settiings.php (these will move to wiki).

> Previously we had all these files sitting in git,

yes but they are all being ignored and not used, so part of this process is to clean up and make a hundreds of duplicated and unused files obsolete.

There is also [settings outside of the repo](https://github.com/govCMS/govcms8lagoon/tree/master/.docker/images/govcms8) which are used sometimes, and these will be made redundant.

Including settings from outside a repo happens on most of the major hosting platforms and this is the pattern I'm shooting for. Pulled this randomly out of one acquia project.

```
  if (!$is_acsf && file_exists('/var/www/site-php') && $site_dir == 'default') {
    require "/var/www/site-php/$ah_group/$ah_group-settings.inc";
  }
```